File: packages/wazeko-transport/src/socket.ts
export interface Socket {
  connect(): Promise<void>;
  send(data: Uint8Array): Promise<void>;
  close(): Promise<void>;
  isConnected(): boolean;
  on(event: "open" | "close" | "error" | "message", listener: (...args: any[]) => void): this;
}

File: crates/wazeko-transport/src/socket.rs
use async_trait::async_trait;
use wazeko_core::error::Result;

#[async_trait]
pub trait Socket: Send + Sync {
    async fn connect(&mut self) -> Result<()>;
    async fn send(&mut self, data: Vec<u8>) -> Result<()>;
    async fn recv(&mut self) -> Result<Option<Vec<u8>>>;
    async fn close(&mut self) -> Result<()>;
    fn is_connected(&self) -> bool;
}
